File: src/php/core/src/Utils/Number.php
<?php

declare(strict_types=1);

namespace DiceBear\Utils;

/**
 * Formats a number for SVG output, rounded to at most 5 decimal places.
 *
 * Rounding to a fixed precision keeps the output bounded and identical across
 * the JS, PHP, and Python ports: every value becomes a multiple of 1e-5 in the
 * SVG coordinate range, which has no exponential form, so the result is built
 * from integer arithmetic — sidestepping PHP's `precision`-based float cast,
 * which would otherwise diverge from the JS reference for small/large values.
 * Five decimals is far below sub-pixel precision for any realistic canvas.
 */
class Number
{
    public static function format(int|float $value): string
    {
        if (is_nan((float) $value)) {
            return 'NaN';
        }

        if (is_infinite((float) $value)) {
            return $value > 0 ? 'Infinity' : '-Infinity';
        }

        $scaled = self::roundHalfUp($value * 100000);
        $sign = $scaled < 0 ? '-' : '';
        $scaled = abs($scaled);

        $integerPart = intdiv($scaled, 100000);
        $fraction = rtrim(str_pad((string) ($scaled % 100000), 5, '0', STR_PAD_LEFT), '0');

        return $sign . $integerPart . ($fraction !== '' ? '.' . $fraction : '');
    }

    /**
     * Rounds half toward +Infinity, matching JS Math.round. Comparing the
     * fractional part against 0.5 — rather than floor($x + 0.5) — avoids the
     * edge case where the largest double below 0.5 would round up to 1.0.
     */
    public static function roundHalfUp(float $value): int
    {
        $floor = floor($value);

        return (int) ($value - $floor < 0.5 ? $floor : $floor + 1);
    }
}

File: src/dart/core/lib/src/utils/number.dart
/// Numeric helpers shared by the PRNG and the renderer: SVG number
/// formatting and the JavaScript-compatible half-up rounding the rest of the
/// engine depends on for cross-language parity.
library;

/// Rounds half toward +∞, matching JavaScript's `Math.round`. Dart's
/// `num.round()` rounds half away from zero, and the naive
/// `(x + 0.5).floor()` over-rounds the largest double below 0.5; comparing
/// the fractional part against 0.5 reproduces `Math.round` exactly.
///
/// Unlike Go (which needs `//go:noinline` here to forbid FMA contraction),
/// Dart rounds every floating-point operation to a 64-bit double on both the
/// VM and the web, so an expression like `value * 100000.0` cannot fuse with
/// the subtraction below. Keep the multiply and the comparison as separate
/// operations anyway — the parity fixture `1.999995 → "2"` is the canary.
double roundHalfUp(double value) {
  final floor = value.floorToDouble();

  if (value - floor < 0.5) {
    return floor;
  }

  return floor + 1.0;
}

/// Formats a number for SVG output, rounded to at most 5 decimal places.
///
/// Rounding to a fixed precision keeps the output bounded and identical across
/// every port: every value becomes a multiple of 1e-5, which has no exponential
/// form, so the result is built from integer arithmetic with no
/// language-specific float stringifying. `double.toString()` must never be used
/// here — Dart renders integral doubles as `1.0` where JavaScript prints `1`.
String formatNumber(double value) {
  if (value.isNaN) {
    return 'NaN';
  }

  if (value == double.infinity) {
    return 'Infinity';
  }

  if (value == double.negativeInfinity) {
    return '-Infinity';
  }

  var scaled = roundHalfUp(value * 100000.0).toInt();
  final sign = scaled < 0 ? '-' : '';

  if (scaled < 0) {
    scaled = -scaled;
  }

  final integerPart = scaled ~/ 100000;
  final fraction = (scaled % 100000)
      .toString()
      .padLeft(5, '0')
      .replaceFirst(_trailingZeros, '');

  if (fraction.isEmpty) {
    return '$sign$integerPart';
  }

  return '$sign$integerPart.$fraction';
}

final _trailingZeros = RegExp(r'0+$');

/// Converts a number to the string JavaScript's `String(value)` produces.
///
/// Used wherever a number becomes a PRNG dedupe/sort key (e.g. `fontWeight`
/// lists), so the key must match the JS port byte for byte: integral values
/// print without a decimal point (`400`, not Dart's `400.0`).
///
/// The zero check comes first on purpose: on the web `-0.0 is int` is true
/// and dart2js prints `-0.0` (where JS prints `0`), so the naive integral
/// branch would silently diverge on the web only.
String jsNumString(num value) {
  if (value == 0) {
    return '0';
  }

  final d = value.toDouble();

  if (d.isNaN) {
    return 'NaN';
  }

  if (d == double.infinity) {
    return 'Infinity';
  }

  if (d == double.negativeInfinity) {
    return '-Infinity';
  }

  if (d == d.truncateToDouble() && d.abs() < 1e21) {
    // BigInt is exact for every integral double; `toInt()` would clamp
    // values at or above 2^63 on the VM, and JS prints e.g. String(1e20)
    // as "100000000000000000000".
    return BigInt.from(d).toString();
  }

  return d.toString();
}
